metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala

```diff
@@ -121,6 +127,182 @@ final class BloopServers(
+      case item
+          if item == Messages.BloopGlobalJsonFilePremodified.useGlobalFile =>
+        Future.successful(())
+    }
+  }
+
+  private def updateBloopGlobalJsonFileThenRestart(
+      bloopGlobalJsonFile: File,
+      requestedBloopJvmProperties: List[String],
+      javaHome: String,
+      reconnect: () => Future[BuildChange]
+  ): Future[Unit] = {
+    writeJVMPropertiesToBloopGlobalJsonFile(
+      bloopGlobalJsonFile,
+      requestedBloopJvmProperties,
+      javaHome
+    ) match {
+      case Failure(exception) => Future.failed(exception)
+      case Success(_) =>
+        languageClient
+          .showMessageRequest(
+            Messages.BloopJvmPropertiesChange.params()
+          )
+          .asScala
+          .flatMap {
+            case messageActionItem
+                if messageActionItem == Messages.BloopJvmPropertiesChange.reconnect =>
+              shutdownServer()
+              reconnect().ignoreValue
+            case _ =>
+              Future.successful(())
+          }
+    }
+
+  }
+
+  private def getBloopGlobalJsonFile(): Option[File] =
+    Try {
+      Paths
+        .get(System.getProperty("user.home"))
+        .resolve(".bloop/bloop.json")
+        .toFile
+    }.toOption
+
+  /**
+   * First we check if the user requested to update the Bloop JVM properties through the extension.
+   * <p>If so, we also check if the Bloop's Global Json file exists and if it was pre-modified by the user.
+   * <p>Then, through consultation with the user through appropriate dialogues we decide if we should
+   * <ul>
+   * <li>overwrite the contents of the Bloop Global Json file with the requested JVM properties and Metal's JavaHome variables; and then restart the Bloop server</li>
+   * <li>or alternatively, leave things untouched</li>
+   * </ul>
+   *
+   * @param maybeRequestedBloopJvmProperties Bloop JVM Properties requested through the Metals Extension settings
+   * @param maybeRunningBloopJvmProperties
+   * @param maybeJavaHome                    Metals' `javaHome`, which Bloop should also preferebly use
+   * @param reconnect                        function to connect back to the build server.
+   * @return `Future.successful` if the purpose is achieved or `Future.failure` if a problem occured such as lacking enough permissions to open or write to files
+   */
+  def ensureDesiredJvmSettings(
+      maybeRequestedBloopJvmProperties: Option[List[String]],
+      maybeRunningBloopJvmProperties: Option[List[String]],
+      maybeJavaHome: Option[String],
+      reconnect: () => Future[BuildChange]
+  ): Future[Unit] = {
+
+    val result = for {
+      requestedBloopJvmProperties <- maybeRequestedBloopJvmProperties
+      javaHome <- maybeJavaHome
+      bloopGlobalJsonFile <- getBloopGlobalJsonFile()
+      bloopGlobalJsonLastModifiedTime <- Try(
+        bloopGlobalJsonFile.lastModified()
+      ).toOption
+      bloopGlobalJsonFileExists <- Try(bloopGlobalJsonFile.exists()).toOption
```

Comment:
```suggestion
      bloopGlobalJsonFileExists = bloopGlobalJsonFile.exists()
```

should be enough. Exists doesn;t throw exceptions

Reply:
this part should be rewritten fundamentally.
The logic of the for comprehension is buggy